chain/gap/find.go

```diff
@@ -0,0 +1,264 @@
+
+	// looks for entriest in the visor processing report table that have been skipped.
+	skipGaps, err := g.findEpochSkips(ctx, maxHeight, g.minHeight)
+	if err != nil {
+		return xerrors.Errorf("detecting skipped gaps: %w", err)
+	}
+	findLog.Infow("found skipped epochs", "count", len(skipGaps))
+
+	var nullRounds visor.ProcessingReportList
+	for _, epoch := range nulls {
+		nullRounds = append(nullRounds, &visor.ProcessingReport{
+			Height:            int64(epoch),
+			StateRoot:         "NULL_ROUND", // let gap fill add the correct state root for this epoch when it runs the consensus task.
+			Reporter:          g.name,
+			Task:              "gap_find",
+			StartedAt:         startTime,
+			CompletedAt:       time.Now(),
+			Status:            visor.ProcessingStatusInfo,
+			StatusInformation: visor.ProcessingStatusInformationNullRound, // the gap finding logic uses this value to exclude null rounds from gap report.
+		})
+	}
+
+	return g.DB.PersistBatch(ctx, skipGaps, heightGaps, taskGaps, nullRounds)
+}
+
+type GapIndexerLens interface {
+	ChainGetTipSetByHeight(ctx context.Context, epoch abi.ChainEpoch, tsk types.TipSetKey) (*types.TipSet, error)
+}
+
+func (g *GapIndexer) findEpochSkips(ctx context.Context, max, min uint64) (visor.GapReportList, error) {
+	log.Debug("finding skipped epochs")
+	reportTime := time.Now()
+
+	var skippedReports []visor.ProcessingReport
+	if err := g.DB.AsORM().ModelContext(ctx, &skippedReports).
+		Order("height desc").
+		Where("status = ?", visor.ProcessingStatusSkip).
+		Where("height >= ?", min).
+		Where("height <= ?", max).
+		Select(); err != nil {
+		return nil, xerrors.Errorf("query processing report skips: %w", err)
+	}
+	log.Debugw("executed find skipped epoch query", "count", len(skippedReports))
+
+	gapReport := make([]*visor.GapReport, len(skippedReports))
+	for idx, r := range skippedReports {
+		gapReport[idx] = &visor.GapReport{
+			Height:     r.Height,
+			Task:       r.Task,
+			Status:     "GAP",
+			Reporter:   g.name,
+			ReportedAt: reportTime,
+		}
+	}
+	return gapReport, nil
+}
+
+func (g *GapIndexer) findEpochGapsAndNullRounds(ctx context.Context, node GapIndexerLens, max, min uint64) (visor.GapReportList, []abi.ChainEpoch, error) {
+	log.Debug("finding epoch gaps and null rounds")
+	reportTime := time.Now()
+
+	var nullRounds []abi.ChainEpoch
+	var missingHeights []uint64
+	res, err := g.DB.AsORM().QueryContext(
+		ctx,
+		&missingHeights,
+		`
+		SELECT s.i AS missing_epoch
+		FROM generate_series(?, ?) s(i)
+		WHERE NOT EXISTS (SELECT 1 FROM visor_processing_reports WHERE height = s.i);
+		`,
+		min, max)
+	if err != nil {
+		return nil, nil, err
+	}
+	log.Debugw("executed find epoch gap query", "count", res.RowsReturned())
+
+	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
+	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
+	for _, gap := range missingHeights {
```

Comment:
Probably want to check whether context is done here so we can shut down nicely.

Reply:
done.